pkg/services/sqlstore/migrations/ualert/channel.go

```diff
@@ -75,245 +110,200 @@ func (m *migration) getNotificationChannelMap() (channelsPerOrg, defaultChannels
-	portedChannels := []*PostableGrafanaReceiver{}
-	var receiver *PostableApiReceiver
+// Create one receiver for every unique notification channel. If there are multiple existing default notification channels, these will be stored in a separate receiver
+// containing them all to be used only on the root-level route. This is to simplify the structure of the migrated routes  as we are currently limited to one receiver per route,
+// so nesting the root-level route will require a significantly more complex setup.
+func (m *migration) createReceivers(amConfig *PostableUserConfig, allChannels []*notificationChannel, defaultChannels []*notificationChannel) (map[uidOrID]*PostableApiReceiver, error) {
+	receivers := make(map[uidOrID]*PostableApiReceiver)
 
-	addChannel := func(c *notificationChannel) error {
-		if c.Type == "hipchat" || c.Type == "sensu" {
-			m.mg.Logger.Error("alert migration error: discontinued notification channel found", "type", c.Type, "name", c.Name, "uid", c.Uid)
-			return nil
-		}
-
-		uid, ok := m.generateChannelUID()
-		if !ok {
-			return errors.New("failed to generate UID for notification channel")
-		}
+	for _, c := range allChannels {
 
-		if _, ok := m.migratedChannelsPerOrg[orgID]; !ok {
-			m.migratedChannelsPerOrg[orgID] = make(map[*notificationChannel]struct{})
-		}
-		m.migratedChannelsPerOrg[orgID][c] = struct{}{}
-		settings, decryptedSecureSettings, err := migrateSettingsToSecureSettings(c.Type, c.Settings, c.SecureSettings)
+		notifier, err := m.createNotifier(c)
 		if err != nil {
-			return err
+			return nil, err
 		}
 
-		portedChannels = append(portedChannels, &PostableGrafanaReceiver{
-			UID:                   uid,
-			Name:                  c.Name,
-			Type:                  c.Type,
-			DisableResolveMessage: c.DisableResolveMessage,
-			Settings:              settings,
-			SecureSettings:        decryptedSecureSettings,
-		})
+		recv := &PostableApiReceiver{
+			Name:                    c.Name, // Channel name is unique within an Org
+			GrafanaManagedReceivers: []*PostableGrafanaReceiver{notifier},
+		}
 
-		return nil
-	}
+		amConfig.AlertmanagerConfig.Receivers = append(amConfig.AlertmanagerConfig.Receivers, recv)
 
-	// Remove obsolete notification channels.
-	filteredChannelUids := make(map[interface{}]struct{})
-	for _, uid := range channelUids {
-		c, ok := allChannels[uid]
-		if ok {
-			// always store the channel UID to prevent duplicates
-			filteredChannelUids[c.Uid] = struct{}{}
-		} else {
-			m.mg.Logger.Warn("ignoring obsolete notification channel", "uid", uid)
+		// Store receivers for creating routes from alert rules later.
+		if c.Uid != "" {
+			receivers[c.Uid] = recv
+		}
+		if c.ID != 0 {
+			// In certain circumstances, the alert rule uses ID instead of uid. So, we add this to be able to lookup by ID in case.
```

Comment:
This is more of a question, but which circumstances? I don't think the ID is guaranteed to be unique within an org. I _think_ based on the way this is used it's fine, since this becomes part of a `map[orgID]map[uidOrID]...` and is already namespaced per-org preventing conflicts. I'm just curious as to why it's needed.

Reply:
I don't remember whether I have this backwards or not... but IIRC one of them is unique per-org, and one is globally unique.

Reply:
Honestly, I was vague here because this is how it was done before and I wasn't able to figure out how it might happen, so left it in to be safe. My best guess is maybe in some older version it could happen and there's the possibility that someone might have some old configs in the database?